src/Pos/TxHistory.hs

```diff
@@ -0,0 +1,241 @@
+  where hasCorrespondingOutput txIn =
+            fmap toBool $ fmap ((== addr) . txOutAddress . toaOut) <$> utxoGet txIn
+        toBool Nothing  = False
+        toBool (Just b) = b
+
+-- | Datatype for returning info about tx history
+data TxHistoryEntry = THEntry
+    { _thTxId       :: !TxId
+    , _thTx         :: !Tx
+    , _thIsOutput   :: !Bool
+    , _thDifficulty :: !(Maybe ChainDifficulty)
+    } deriving (Show, Eq, Generic)
+
+makeLenses ''TxHistoryEntry
+
+-- | Type of monad used to deduce history
+type TxSelectorT m = UtxoStateT (MaybeT m)
+
+-- | Select transactions related to given address. `Bool` indicates
+-- whether the transaction is outgoing (i. e. is sent from given address)
+getRelatedTxs
+    :: Monad m
+    => Address
+    -> [(WithHash Tx, TxWitness, TxDistribution)]
+    -> TxSelectorT m [TxHistoryEntry]
+getRelatedTxs addr txs = fmap DL.toList $
+    lift (MaybeT $ return $ topsortTxs (view _1) txs) >>=
+    foldlM step DL.empty
+  where
+    step ls (WithHash tx txId, _wit, dist) = do
+        let isIncoming = tx `hasReceiver` addr
+        isOutgoing <- tx `hasSender` addr
+        let allToAddr = all ((== addr) . txOutAddress) $ _txOutputs tx
+            isToItself = isOutgoing && allToAddr
+        if isOutgoing || isIncoming
+            then do
+            applyTxToUtxo (WithHash tx txId) dist
+            identity %= filterUtxoByAddr addr
+
+            -- Workaround to present A to A transactions as a pair of self-canceling
+            -- transactions in history
+            let resEntry = THEntry txId tx isOutgoing Nothing
+                resList = if isToItself
+                          then DL.fromList [resEntry & thIsOutput .~ False, resEntry]
+                          else DL.singleton resEntry
+            return $ ls <> resList
+            else return ls
+
+-- | Given a full blockchain, derive address history and Utxo
+-- TODO: Such functionality will still be useful for merging
+-- blockchains when wallet state is ready, but some metadata for
+-- Tx will be required.
+deriveAddrHistory
+    -- :: (Monad m, Ssc ssc) => Address -> [Block ssc] -> TxSelectorT m [TxHistoryEntry]
+    :: (Monad m) => Address -> [Block ssc] -> TxSelectorT m [TxHistoryEntry]
+deriveAddrHistory addr chain = identity %= filterUtxoByAddr addr >>
+                               deriveAddrHistoryPartial [] addr chain
+
+deriveAddrHistoryPartial
+    :: (Monad m)
+    => [TxHistoryEntry]
+    -> Address
+    -> [Block ssc]
+    -> TxSelectorT m [TxHistoryEntry]
+deriveAddrHistoryPartial hist addr chain =
+    DL.toList <$> foldrM updateAll (DL.fromList hist) chain
+  where
+    updateAll (Left _) hst = pure hst
+    updateAll (Right blk) hst = do
+        txs <- getRelatedTxs addr $
+                   map (over _1 withHash) (blk ^. blockTxas)
+        let difficulty = blk ^. difficultyL
+            txs' = map (thDifficulty .~ Just difficulty) txs
```

Comment:
Sure, but I suspect that not everyone is aware of `?~` and it's not even that useful to force its usage